som/display.go

```diff
@@ -0,0 +1,120 @@
+package som
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/gonum/matrix/mat64"
+)
+
+// Creates an SVG representation of the U-Matrix of the given codebook.
+// codebook is the codebook we're displaying the U-Matrix for,
+// coordsDims are the dimensions of the grid,
+// uShape is the shape of the grid,
+// title is the title of the output SVG,
+// and writer is the io.Writter to write the output SVG to.
+func UMatrixSVG(codebook *mat64.Dense, coordsDims []int, uShape string, title string, writer io.Writer) error {
+	xmlEncoder := xml.NewEncoder(writer)
+	// array to hold the xml elements
+	elems := []interface{}{h1{Title: title}}
+
+	rows, _ := codebook.Dims()
+	distMat, _ := DistanceMx("euclidean", codebook)
+	distMatRows, distMatCols := distMat.Dims()
+	coords, _ := GridCoords(uShape, coordsDims)
+	coordsDistMat, _ := DistanceMx("euclidean", coords)
+
+	// get maximum distance between codebook vectors
+	// maximum distance will be rgb(0,0,0)
+	MAX := 0.0
+	for i := 0; i < distMatRows; i++ {
+		for j := i; j < distMatCols; j++ {
+			if distMat.At(i, j) > MAX {
+				MAX = distMat.At(i, j)
+			}
+		}
+	}
+	MUL := 20.0
+	OFF := 10.0
+	// function to scale the coord grid to something visible
+	scale := func(x float64) float64 { return MUL*x + OFF }
+
+	svgElem := svgElement{Width: float64(coordsDims[1])*MUL + 2*OFF, Height: float64(coordsDims[0])*MUL + 2*OFF, Polygons: make([]polygon, rows)}
+	elems = append(elems, svgElem)
+	for row := 0; row < rows; row++ {
+		coord := coords.RowView(row)
+		cbVec := codebook.RowView(row)
+		avgDistance := 0.0
+		// this is a rough approximation of the notion of neighbor grid coords
+		allRowsInRadius := allRowsInRadius(row, math.Sqrt2*1.01, coordsDistMat)
+		for _, rwd := range allRowsInRadius {
+			if rwd.Dist > 0.0 {
+				otherMu := codebook.RowView(rwd.Row)
+				cbvDist, _ := Distance("euclidean", cbVec, otherMu)
+				avgDistance += cbvDist
+			}
+		}
+		avgDistance /= float64(len(allRowsInRadius) - 1)
+		color := int((1.0 - avgDistance/MAX) * 255.0)
+		polygonCoords := ""
+		x := scale(coord.At(0, 0))
+		y := scale(coord.At(1, 0))
+		xOffset := 0.5 * MUL
+		yOffset := 0.5 * MUL
+		// hexagon has a different yOffset
+		if uShape == "hexagon" {
+			yOffset = math.Sqrt(0.75) / 2.0 * MUL
+		}
+		// draw a box around the current coord
+		polygonCoords += fmt.Sprintf("%f,%f ", x+xOffset, y+yOffset)
+		polygonCoords += fmt.Sprintf("%f,%f ", x+xOffset, y-yOffset)
+		polygonCoords += fmt.Sprintf("%f,%f ", x-xOffset, y-yOffset)
+		polygonCoords += fmt.Sprintf("%f,%f ", x-xOffset, y+yOffset)
+		polygonCoords += fmt.Sprintf("%f,%f ", x+xOffset, y+yOffset)
+
+		svgElem.Polygons[row] = polygon{
+			Points: []byte(polygonCoords),
+			Style:  fmt.Sprintf("fill:rgb(%d,%d,%d);stroke:black;stroke-width:1", color, color, color),
+		}
+	}
+
+	xmlEncoder.Encode(elems)
+	xmlEncoder.Flush()
+
+	return nil
+}
+
+type rowWithDist struct {
```

Comment:
Normally you should define these on top of the src file or right in front of the `func` which uses these structures.